File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> l;
    public List<Integer> ll;
    public HashMap<Integer, Integer> hm;
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        l = new ArrayList<>();
        ll = new ArrayList<>();
        // Arrays.sort(candidates);
        // hm = new HashMap<>();
        // for (int i = 0;i<candidates.length;i++)
        // {
        //     hm.put(candidates[i], i);
        // }
        doCombinations(candidates, target, 0, 0);
        return l;
    }
    
    public void doCombinations(int[] candidates, int target, int i, int sum)
    {
        if (sum == target)
        {
            l.add(new ArrayList<>(ll));
            return;
        }
        if (i == candidates.length || sum > target)
        {
            return;
        }
        ll.add(candidates[i]);
        doCombinations(candidates, target, i, sum + candidates[i]);
        
        ll.remove(ll.size() - 1);
        doCombinations(candidates, target, i+1, sum);
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        temp = []
        def dp(i, sum):
            if i >= len(candidates) or sum > target:
                return
            if sum == target:
                res.append(temp.copy())
                return
            sum += candidates[i]
            temp.append(candidates[i])
            dp(i, sum)
            sum -= candidates[i]
            temp.pop()
            dp(i+1, sum)
        dp(0, 0)
        return res
        